### ISSUE TITLE
ENH `GeneralizingEstimator` could be faster if batched

### ISSUE THREAD

**Issue by mathias-sm** [mne-tools/mne-python#13906]
### Describe the new feature or enhancement

Hi!

To the best of my understanding, the current implementation of `GeneralizingEstimator` performs a nested loop: for each timepoint, `fit`, then for each timepoint, `transform`+`score` the estimator [1]. However, once the estimator is fitted, it is possible to batch the `transform` instead of calling it once per timepoint by reshaping along that axis and then reshaping the other way around. I have found this to be much faster in practice, see below.

[1] I'll confess I do not fully understand [the implementation of `GeneralizingEstimator._transform`](https://github.com/mne-tools/mne-python/blob/31219b44380b0429dad2a143b865a2bbe3f6f6f4/mne/decoding/search_light.py#L505-L536) yet...

### Describe your proposed implementation

Provided a better understanding of `GeneralizingEstimator._transform` it should be possible to follow the logic below.

Given a classifier `clf`, training time `t_train` and `X` of dimensions `(n_trials, n_channels, n_times)`:
1. Fit the classifier:
`clf.fit(X[:, :, t_train], y)`
2. Flatten the test data along the `(n_trials, n_times)` dimensions:
`X_test_flat = X.transpose(0, 2, 1).reshape(-1, n_channels)`
3. Classify, then compare to labels that have been repeated along the time dimension:
`correct = clf.predict(X_test_flat) == np.repeat(y, n_times)`
4. Reshape back to `(n_trials, n_times)` and average along `(n_trials)`:
`scores.append(np.mean(correct.reshape(len(test_idx), n_times)))`

I propose a MWE + a comparison below, between the current `GeneralizingEstimator`, nested loops, and the batched version:

```python
import time
import numpy as np
import mne
import sklearn

n_trials, n_channels, n_times = 100, 100, 100
X = np.random.randn(n_trials, n_channels, n_times)
y = np.random.randint(0, 2, n_trials)
clf = sklearn.linear_model.LogisticRegression()
cv = list(sklearn.model_selection.KFold(n_splits=5).split(X))

ts = [time.time()]
time_gen = mne.decoding.GeneralizingEstimator(clf, n_jobs=None, scoring="accuracy", verbose=False)
mne_scores = mne.decoding.cross_val_multiscore(time_gen, X, y, cv=cv, n_jobs=None, verbose=False).mean(0)

ts.append(time.time())
loop_scores = np.zeros((len(cv), n_times, n_times))
for k, (train_idx, test_idx) in enumerate(cv):
    for t_train in range(n_times):
        clf.fit(X[train_idx, :, t_train], y[train_idx])
        for t_test in range(n_times):
            loop_scores[k, t_train, t_test] = (clf.predict(X[test_idx, :, t_test]) == y[test_idx]).mean()
loop_scores = loop_scores.mean(0)

ts.append(time.time())
batch_scores = np.zeros((len(cv), n_times, n_times))
for k, (train_idx, test_idx) in enumerate(cv):
    for t_train in range(n_times):
        clf.fit(X[train_idx, :, t_train], y[train_idx])
        X_test_flat = X[test_idx].transpose(0, 2, 1).reshape(-1, n_channels)
        correct = clf.predict(X_test_flat) == np.repeat(y[test_idx], n_times)
        batch_scores[k, t_train] = correct.reshape(len(test_idx), n_times).mean(0)
batch_scores = batch_scores.mean(0)
ts.append(time.time())

for name, dt in zip(["GeneralizingEstimator", "loop", "batched"], np.diff(ts)):
    print(f"{name}: {dt:.2f}s")

print(f"{np.allclose(loop_scores,  mne_scores)=}")
print(f"{np.allclose(batch_scores, mne_scores)=}")
```

Which outputs:

```
GeneralizingEstimator: 10.26s
loop: 2.09s
batched: 1.03s
np.allclose(loop_scores,  mne_scores)=True
np.allclose(batch_scores, mne_scores)=True
```

I understand the sliding/generalizing estimators are embarassingly parallelizeable, but on real data, across subjects, etc. I have found this to be substantially faster. I do not know why there is a difference between a naive nested loop the current implementation, but this suggests to me that this MWE is nowhere near as general as `GeneralizingEstimator`, see below.

Just to get a feel for the scaling, I ran the MWE above with `n_times` ranging from `10` to `250` and there's almost a factor 100 between the coefficients in front of a quadratic terms (I assume one can't do better than O(n_times²) given the problem).

[Code to generate this here](https://github.com/user-attachments/files/28073758/benchmark_gat_plot.py), output:

<img width="755" height="401" alt="Image" src="https://github.com/user-attachments/assets/bd2e2ae9-b1ab-4502-aa30-5170f4d82efe" />

I think this could be implemented instead of the present implementation in a way that is numerically identical.

### Describe possible alternatives

It is entirely possible that the implementation I propose loses generality compared to the existing one, and I have simply not quite understood the current implementation yet.

With that said, reasoning about the new version would be relatively straightforward. Given the speed increase in a simple case, even if there is a loss in generality, it might be worth implementing both and something that routes to one implementation or the other.

Assuming the quadratic scaling law is a good approximation, and given a similar number of trials, sensors, and cv-folds: 600Hz over 2 seconds, estimation goes from 6 minutes to 10 seconds.

It seems to be that:
1. The higher `n_times`, the more visible the improvement is.
2. The higher `n_trials` or `n_channels` are, the less visible the improvement is.

### Additional context

Thanks everyone!

**Comment by agramfort** [comment]
yes you could reshape the data to transform many time points in parallel. I
think it's what you are suggesting

**Comment by mathias-sm** [comment]
Yes, this is what I am suggesting. I do this in a project of mine "manually", but I was thinking that `mne`'s `GeneralizingEstimator` could benefit from it too.

If it seems worth trying, I'd be happy to draft a PR at some point.

**Comment by agramfort** [comment]
well you do see a big benefit from it so it seems relevant :)

**Comment by mathias-sm** [comment]
I was wondering if I have missed something like my solution working in a simple case but not being as general.

At some point I'll draft a PR. I see there are GAT tests in `decoding/tests/test_base.py` and in `decoding/tests/test_search_light.py`: if they still pass, I'll report about speed benefits if any :)

**Comment by mathias-sm** [comment]
PR #13909  was merged – I'd be curious to see if this has any visible impact in some pipeline (e.g. https://github.com/mne-tools/mne-bids-pipeline has 5 examples datasets where GAT is performed).
But in any case, this issue can be closed now, thanks everyone for being patient!

**Comment by larsoner** [comment]
Looks like runs of ERP_CORE_ERN for mne-bids-pipeline for time gen step [now take](https://app.circleci.com/pipelines/github/mne-tools/mne-bids-pipeline/5675/workflows/b96a4975-c414-46fd-87e2-2b0148c4c05e/jobs/108448) ~51s, whereas [a week ago](https://app.circleci.com/pipelines/github/mne-tools/mne-bids-pipeline/5667/workflows/fef5dbb0-48da-4bb1-ab1b-5462672bfa8d/jobs/108085) they took ~2m11s. So > 2x speedup there! 🚀
